### DISCUSSION TITLE
Application to Vertical Pitches

### DISCUSSION THREAD

**Discussion by larsmaurath** [DonsetPG/narya#23] [General]
As mentioned in other threads, I have started generating training data with the goal of applying the narya models to vertical pitches. 

Here I want to share the data, trained weights and report on the initial experience of applying this to a different set of data.

## Data

Images should already be in the correct format to apply the trainer as is. (512x512 for the direct homography, 320x320 for the keypoints estimation). The format of the homography matrix and the keypoints are also modeled after the orginal data set.

### Direct Homography

976 samples across multiple matches, cameras and pitch conditions.

[Samples](https://www.dropbox.com/s/ykiezvylr1zwj7m/samples_direct_homography.zip?dl=0)

Weights are trained with 50 epochs for now.

[Weights](https://www.dropbox.com/s/8t98kx7ezsxazxk/HomographyModel_0.0001_32.h5?dl=0)

### Keypoints

579 samples across multiple matches, cameras and pitch conditions.

[Samples](https://www.dropbox.com/s/q6ka08h5wicfmm8/samples_keypoints.zip?dl=0)

Weights are trained with 50 epochs for now.

[Weights](https://www.dropbox.com/s/q2pv4dyn69bqtvv/FPN_efficientnetb3_0.0001_32.h5?dl=0)

### Direct Homography

My experience has been that the results for the direct homography have been disappointing for vertical pitches. @DonsetPG has already shed some light [here](https://github.com/DonsetPG/narya/issues/6) on what the drivers of this could be.

Some examples of results I get for the trained model above:

<img width="311" alt="direct_homography_1" src="https://user-images.githubusercontent.com/11316760/110189677-da697e80-7ded-11eb-9cca-318fa0e5f5b4.png">
<img width="311" alt="direct_homography_2" src="https://user-images.githubusercontent.com/11316760/110189678-db021500-7ded-11eb-972e-fc06392cb951.png">
<img width="310" alt="direct_homography_3" src="https://user-images.githubusercontent.com/11316760/110189679-db021500-7ded-11eb-9550-b8c0e36af46e.png">

### Keypoints

I had much more success with the keypoints model. My intuition would also be that this should work better for vertical pitches than for the default broadcast camera as you often see many more keypoints per frame. I'd expect that player tracking will be harder in the top half of the pitch when the angle gets very flat.

Some examples of results I get for the trained model above:

<img width="313" alt="keypoints_1" src="https://user-images.githubusercontent.com/11316760/110189889-ce31f100-7dee-11eb-83c2-518e14c70796.png">
<img width="310" alt="keypoints_2" src="https://user-images.githubusercontent.com/11316760/110189890-ceca8780-7dee-11eb-9949-38d8d8f2f6ed.png">
<img width="311" alt="keypoints_3" src="https://user-images.githubusercontent.com/11316760/110189892-ceca8780-7dee-11eb-8d56-adad07a965cc.png">
<img width="308" alt="keypoints_4" src="https://user-images.githubusercontent.com/11316760/110189893-ceca8780-7dee-11eb-934f-af0e2228de8f.png">

## Next Steps

I am currently working on generating training data for the player tracking and will also share once successful.

**Comment by DonsetPG** [comment]
Hey,

Thanks a lot for sharing this looks great! 
I took the liberty of including the weights and datasets in the same GCP bucket:

* [Vertical Homography Dataset](https://storage.googleapis.com/narya-bucket-1/dataset/vertical_samples_direct_homography.zip)
* [Vertical Homography Model](https://storage.googleapis.com/narya-bucket-1/models/vertical_HomographyModel_0.0001_32.h5)
* [Vertical Keypoints Model](https://storage.googleapis.com/narya-bucket-1/models/vertical_FPN_efficientnetb3_0.0001_32.h5)
* [Vertical Keypoints Dataset](https://storage.googleapis.com/narya-bucket-1/dataset/vertical_samples_keypoints.zip)

I hope that is ok with you. 

Would it be possible for you to make a pull request to add:
- the models/weights in the readme with the others? 
- add a link to your annotation tool? 

Again, thanks a lot!
Cheers

**Reply by larsmaurath** [reply]
Of course, happy to share them on the readme. I'll declare them as experimental for now, and I'll update them once I have more samples.

Also working on cleaning up the annotation tool. Added the player labeling last week. Will add once I have it add on Github.

**Comment by ShanglinMo** [comment]
Hello larsmaurah, 

Thank you very much for sharing the vertical pitch model! 

I'm trying to apply the model to a vertical pitch, the key points look good, but the wrapping part goes wrong. I wonder if you may know what I missed? Any advice would be appreciated. Thank you very much!
![image](https://user-images.githubusercontent.com/33772538/127680540-629d5806-6919-47d3-85d7-cacd4fb574d8.png)

**Reply by larsmaurath** [reply]
Hey @ShanglinMo, sorry for the late reply. I also have issues with getting the warping right for your example (see below). Mine seems to have issues with finding the right key points, while I agree though that your keypoints look correct.
![test_00000](https://user-images.githubusercontent.com/11316760/130329976-24ced3cd-2609-4ad7-8cac-44ad92fa3068.jpg)

**Reply by larsmaurath** [reply]
Could you check if your code works with the below (easier) example? Below works for me.

![frame1](https://user-images.githubusercontent.com/11316760/130330003-52146362-a00a-4bf8-9b8d-bd858c4d6fb9.jpg)

**Reply by ShanglinMo** [reply]
Hello @larsmaurath, thank you for the test, your example works well for me. 

About my example, I find that the coordinate of the keypoints are fine, but some of them are not correctly labeled. I prepare to use the tool you mentioned to create my dataset to retrain the model. 

Thank you very much again for all your help :)

![image](https://user-images.githubusercontent.com/33772538/130332873-551f553f-0eed-4174-b9af-ff3a287934d5.png)
